File: src/action/RedAction.java
package action;
import game.Player;

public class RedAction extends Action {
    public RedAction(int holeNumber, Player player) {
        super(holeNumber, player);
    }

    @Override
    public int execute(int [][] board) {
        int nbSeed = board[holeNumber][1];
        board[holeNumber][1] = 0;

        if (nbSeed != 0) {
            int i = holeNumber;
            for (; nbSeed > 0; i = (i + 1) % 16)
                if (i != holeNumber) {
                    board[i][1]++;
                    nbSeed--;
                }
            i = (i == 0) ? 15 : i - 1;
            return seedCapturing(i, board);
        }
        return 0;
    }

    @Override
    public String toString() {
        return super.toString() + "R";
    }
}

File: projectC/awale/src/action/redAction.c
#include "../../headers/action/redAction.h"

RedAction createRedAction(int holeNumber, Player player)
{
    RedAction redAction;
    redAction.action.holeNumber = holeNumber;
    redAction.player = player;

    return redAction;
}

int execute_redAction(RedAction *redAction, int board[16][3])
{
    int nbSeed = board[redAction->action.holeNumber][1];
    board[redAction->action.holeNumber][1] = 0;

    if (nbSeed != 0)
    {
        int i = redAction->action.holeNumber;

        for (; nbSeed > 0; i = (i + 1) % 16)
        {
            if (i != redAction->action.holeNumber)
            {
                board[i][1]++;
                nbSeed--;
            }
        }

        i = (i == 0) ? 15 : i - 1;

        return seedCapturing(i, board);
    }

    return 0;
}

char *toStringRedAction(RedAction *redAction)
{
    char *actionStr = toStringAction(&(redAction->action));
    char *result = malloc(strlen(actionStr) + 2);
    strcpy(result, actionStr);
    strcat(result, "R");
    free(actionStr);
    return result;
}
